### ISSUE TITLE
hasListener(event)

### ISSUE THREAD

**Issue by drkibitz** [primus/eventemitter3#26]
This was something I figured would be a good time to ask for as well since you are probably working hard on the next major version with the previous enhancement just implemented.

So what I am looking for is something fast and simple that lets me know if there is > 0 listeners for a particular event.

Here is what I've done so far, be kind ;)
https://github.com/GoodBoyDigital/pixi.js/blob/dev/src/ticker/Ticker.js#L6-L18

Basically, the `listeners` method was just too much overhead for what I would be using it for, and I could keep the count myself, but that get's more difficult with `once` method.

May I please have your thoughts?

**Comment by drkibitz** [comment]
I know the `_events` object is private, but it's the same for all the different EventEmitter implementations, so that why I did that check, at this point it feels more like a "use at your own risk" property.

**Comment by drkibitz** [comment]
Maybe a helper static method would be an ok solution since it wouldn't pollute the EventEmitter prototype, but also be safe to access `_event` from within the same module.

```javascript
EventEmitter.hasListener = function (emitter, event) {
    if (emitter._events) {
        if (event) return !!emitter._events[event];
        return true;
    }
    return false;
};
```

**Comment by 3rd-Eden** [comment]
While I understand that there is a use-case for this, I don't think it belongs in the API of an EventEmitter because there is already an API that provides this information, which is the `listeners` method. Yes it introduces some overhead by returning all the listeners.

FYI: your implementation does not seem to take the prefixing of keys in to account. They event names are prefixed with `~` to prevent object based dictionary attacks so event names such as __proto__ or toString will not completely destroy the EventEmitter.

**Comment by drkibitz** [comment]
@3rd-Eden Thanks for the info on the prefix, though when I looked at it again, it's working because it's a composed emitter that only supports 1 event, so the `_events` check is all it really needs there.

As for using the `listeners` method, I hope you will reconsider because the overhead it introduces is unacceptable for my use case. It returns a new object every single time it is called, and we are using your library within a rendering engine that touts itself as *fast*, which you can conclude is the reason why we are using your library. We would not want more allocations by simple listener check. Sometimes there's a loop, but there is a new object every time.

In the rendering engine, the listener check itself is a performance optimization in order to not request animation frames if nothing is listening. So if the performance optimization has overhead, we're doing it wrong.

Regardless of all this, I'll still just change my function to this ;)
```javascript
function hasListeners(emitter)
{
    return !!emitter._events;
}

**Comment by drkibitz** [comment]
@3rd-Eden Also I couldn't quite gauge your response in terms of what you think of actually external apis referencing `_events`. I got that you believe this doesn't belong in an EventEmitter api, but what I didn't get was where it did belong. I hope after my last comment you may have a better view of the user case, and can suggest that I keep or lose the reference to `_events`.

**Comment by 3rd-Eden** [comment]
What I could do as compromise is add a boolean to the `listeners` method which just checks if the listeners exists for the given event instead returning an array. This way we don't introduce any new API's but still get the desired performance boost.

**Comment by drkibitz** [comment]
@3rd-Eden that sounds good. If I understand correctly, and boolean argument, which will change the return value to a boolean?

**Comment by 3rd-Eden** [comment]
@drkibitz Exactly. 

```js
EE = new EventEmitter();

EE.on('event', function () {});
EE.listeners('event', true); // true 
EE.listeners('unknown', true); // false 
```

**Comment by drkibitz** [comment]
Great, appreciate the compromise. Will look forward to see the implementation, need any help?

**Comment by 3rd-Eden** [comment]
@drkibitz Already done, writing tests as we speak :)

**Comment by drkibitz** [comment]
@3rd-Eden your awesome.

**Comment by 3rd-Eden** [comment]
Landed in the newly released `1.0.0` build
